R/render-plot.R

```diff
@@ -266,26 +315,366 @@ drawPlot <- function(name, session, func, width, height, pixelratio, res, ...) {
+    colour <- geom$default_aes$colour
+    fill <- geom$default_aes$fill
+    # To avoid the possibility of modifying twice
+    if (identical(colour, default_color)) {
+      geom$default_aes$colour <- adjust_color(colour, bg, fg, accent)
+    }
+    if (identical(fill, default_fill)) {
+      geom$default_aes$fill <- adjust_color(fill, bg, fg, accent)
+    }
+  }, geoms, default_colours, default_fills)
+
+  # Restore defaults
+  on.exit({
+    Map(function(geom, colour, fill) {
+      geom$default_aes$colour <- colour
+      geom$default_aes$fill <- fill
+    }, geoms, default_colours, default_fills)
+  }, add = TRUE)
+
+  # Set scaled aes defaults (if appropriate)
+  p <- add_scale_defaults(p, "colour", theme)
+  p <- add_scale_defaults(p, "fill", theme)
 
-  gtable <- ggplot2::ggplot_gtable(build)
-  grid::grid.draw(gtable)
+  if (newpage) grid::grid.newpage()
 
-  structure(list(
-    build = build,
-    gtable = gtable
-  ), class = "ggplot_build_gtable")
+  ggplot_build(p)
+}
+
+ggtheme_auto <- function(bg, fg) {
+  text <- ggplot2::element_text(colour = fg)
+  line <- ggplot2::element_line(colour = fg)
+  themeGray <- ggplot2::theme_gray()
+
+  ggplot2::theme(
+    line = line,
+    text = text,
+    axis.title = text,
+    axis.text = text,
+    axis.ticks = line,
+    plot.background = ggplot2::element_rect(fill = bg, colour = "transparent"),
+    panel.background = ggplot2::element_rect(
+      fill = adjust_color(themeGray$panel.background$fill, bg, fg)
+    ),
+    panel.grid = ggplot2::element_line(colour = bg),
+    legend.background = ggplot2::element_rect(fill = "transparent"),
+    legend.box.background = ggplot2::element_rect(
+      fill = "transparent", colour = "transparent"
+    ),
+    legend.key = ggplot2::element_rect(
+      fill = adjust_color(themeGray$legend.key$fill, bg, fg),
+      colour = bg
+    ),
+    strip.background = ggplot2::element_rect(
+      fill = adjust_color(themeGray$strip.background$fill, bg, fg)
+    ),
+    strip.text = text
+  )
+}
+
+# Logic for adjusting a color based on bg/fg/accent
+adjust_color <- function(color, bg, fg, accent = NA) {
+  if (!length(color)) return(color)
+  if (length(color) > 1) {
+    warning("Failed to translated aes defaults (expected to be of length 1)")
+    return(color)
+  }
+  if (is.na(color) || identical(color, "NA")) return(color)
+
+  # If a gray scale color, then the degree of gray determines
+  # the mixing between fg (aka black) and bg (aka white)
+  rgbs <- grDevices::col2rgb(color, alpha = TRUE)[1:3,1]
+  if (sum(diff(rgbs)) == 0) {
+    return(mix_colors(bg, fg, 1 - (rgbs[1] / 255)))
+  }
+
+  # At this point we should be dealing with an accent color...
+  # If accent is NA though, then the user has specified to NOT change it
+  if (is.na(accent)) color else accent
+}
+
+
+mix_colors <- function(bg, fg, amount) {
+  if (!length(bg) || !length(fg)) return(NULL)
+  mid_color <- grDevices::colorRamp(c(bg, fg), alpha = TRUE)(amount)
```

Comment:
Look into whether we can/should do better than a linear interpolation between these two colors

Reply:
FWIW, for the `bs_theme_base_colors` impl, I tried switching to `space="Lab"` and it immediately complained about `alpha=TRUE`, that was enough to make me leave it as linear.